### ISSUE TITLE
I have a question about whether Hungarian matching is essential for semantic segmentation

### ISSUE THREAD

**Issue by alexaex** [tue-mps/eomt#36]
Most query-based detectors or segmenters, such as DETR, Mask2Former, and your awesome work EoMT(Encoder Only Masked Transformer), use a set of queries to interact with image features and predict masks or bounding boxes.

To the best of my knowledge, Hungarian matching assigns each query to a ground-truth instance, enabling the same architecture to handle instance, semantic, and panoptic segmentation.

But for semantic segmentation, is hungarian matching necessary?
Suppose we have query embeddings Q of shape [N, Q, D] and image features X of shape [N, L, D]. We can predict semantic segmentation results and train the model without complex Hungarian matching using the following steps:
```
X = X.transpose(1, 2).continous().view(N, -1, L//grid_size, L//grid_size)
X = upsample(X) # [N, D, H/4, W/4] 1/4 scale
mask_embedding = MLP(Q) # [N, Q, D] -> [N, Q, D]

mask_logits = torch.einsum('nqd, ndhw->ndhw', mask_embedding, X)
class_embedding = MLP(Q) # [N, Q, D] -> [N, Q, num_classes+1]

mask_logits = torch.einsum('nqc, nqhw->nchw', class_embedding.softmax(dim=-1), mask_logits.sigmoid())

**Comment by NiccoloCavagnero** [comment]
Hi @alexaex,

interesting question. Indeed you don't strictly need Hungarian matching to perform semantic segmentation, but it can be even simpler than how you are posing it. 

If you want to keep using queries, simply having `Q = number_of_classes` would do the job. Alternatively, an even simpler linear layer on top of the image features would work too.

Best,
Nick

**Comment by alexaex** [comment]
> Hi [@alexaex](https://github.com/alexaex),
> 
> interesting question. Indeed you don't strictly need Hungarian matching to perform semantic segmentation, but it can be even simpler than how you are posing it.
> 
> If you want to keep using queries, simply having `Q = number_of_classes` would do the job. Alternatively, an even simpler linear layer on top of the image features would work too.
> 
> Best, Nick

Great idea! Thanks for your reply.

I wanna skip the hungarian matching because many of medical images contain only background. This causes the 'targets' to become  ‘None', so the mask2former loss can't be computed due to the lack of supervision. In general, the pixel classification modeling doesn't have this trouble.  To handle this, I think I should drop the image without any foreground or fall back to pixel classification.

Anyway, thanks for your great answer!
